db/virtual_table.cc

```diff
@@ -108,10 +97,25 @@ mutation_source memtable_filling_virtual_table::as_mutation_source() {
         };
 
         auto qr = make_lw_shared<my_query_restrictions>(range);
-        auto f = execute(*mt, *qr).then([qr] {});
+
+        auto populate = [this, qr, s, permit, range, slice, pc, trace_state, fwd, fwd_mr] {
+            auto mt = make_lw_shared<memtable>(schema());
+
+            return execute(*mt, *qr).then([qr] {}).then([this, mt, s, permit, range, slice, pc, trace_state, fwd, fwd_mr] () {
+                auto rd = make_restricted_flat_reader(mt->as_data_source(), s, permit, range, slice, pc, trace_state, fwd, fwd_mr);
```

Comment:
make_restricted_flat_reader takes range by reference and assumes it will be alive during whole read, but it lives inside lambda captures so will be destroyed immediately.

Take the range from qr, which lives long enough.